### PR TITLE
V1.2 Substring-Based 'Filter' Command

### LINKED ISSUE #68: Implement Filter by Name
User can filter clients by name.

### LINKED ISSUE #88: Update find feature to search substring in names instead of equals

### PR DESCRIPTION
**Pull request by jyztintan**
This PR builds from the AB3's original keyword search functionality 'find' command to a substring-based keyphrase 'filter' command.

# Original Implementation
The previous implementation of 'find' command searches for any contacts whose names match with any of the keywords input. 
> In the original implementation, "find James Udhaya" returns all contacts with 'James' or 'Udhaya' as a word in their name.

# Key changes include
- Transition to 'Filter' Command: Renamed all instances of the 'find' command to 'filter'
- The search keyphrase _need not be an **exact word**_ in the contact name, allowing for more flexible searches.
- The command no longer supports multiple keyphrases/keywords; it focuses on a single substring match.
- Modified testing to reflect the new expected behaviour of the 'filter' command.
- Refinement of Search Parameters: The terminology within the application has been updated from 'keyword' to 'keyphrase' to better reflect the nature of the substring search, emphasizing that searches are not limited to whole words.

> In the new implementation, "filter James Udhaya" only shows contacts with "James Udhaya" in their name, for e.g. "James Udhayaaaaa Tan"

# Notably, the following has been kept consistent
- Case-Insensitive Searching: Updated the search mechanism to be case-insensitive, allowing users to search without worrying about the case of the letters, further enhancing the user experience.
- Resultant contacts are still reflected as a new indexed list
- Filter commands with empty parameter are not allowed

# Rationale
The shift from a keyword-based search to a substring-based 'filter' command is designed to offer a more flexible and user-centric search experience. Recognizing that users may not always remember the exact spelling or full names of their contacts, the new 'filter' command accommodates partial and varied inputs, making the search process more inclusive and less prone to error.

Fix #68, fix #88

### CODE REVIEW

**Comment by UdhayaShan1** [comment]
Smoke tests seem find so far
`filter` returns invalid command
`filter 213`  returns 0
`filter ud` returns 1 Ud profile

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 1]
```diff
@@ -134,8 +133,8 @@ public static void showPersonAtIndex(Model model, Index targetIndex) {
         assertTrue(targetIndex.getZeroBased() < model.getFilteredPersonList().size());
 
         Person person = model.getFilteredPersonList().get(targetIndex.getZeroBased());
-        final String[] splitName = person.getName().fullName.split("\\s+");
-        model.updateFilteredPersonList(new NameContainsKeywordsPredicate(Arrays.asList(splitName[0])));
+        final String fullName = person.getName().fullName;
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java:136]: 👍

File: src/test/java/seedu/address/logic/parser/AddressBookParserTest.java [hunk 2]
```diff
@@ -69,11 +65,11 @@ public void parseCommand_exit() throws Exception {
     }
 
     @Test
-    public void parseCommand_find() throws Exception {
-        List<String> keywords = Arrays.asList("foo", "bar", "baz");
-        FindCommand command = (FindCommand) parser.parseCommand(
-                FindCommand.COMMAND_WORD + " " + keywords.stream().collect(Collectors.joining(" ")));
-        assertEquals(new FindCommand(new NameContainsKeywordsPredicate(keywords)), command);
+    public void parseCommand_filter() throws Exception {
+        String keyphrase = "foo bar baz";
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/logic/parser/AddressBookParserTest.java:69]: 👍

File: src/test/java/seedu/address/logic/commands/FilterCommandTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    public void toStringMethod() {
+        NameContainsKeyphrasePredicate predicate = new NameContainsKeyphrasePredicate("keyphrase");
+        FilterCommand filterCommand = new FilterCommand(predicate);
+        String expected = FilterCommand.class.getCanonicalName() + "{predicate=" + predicate + "}";
+        assertEquals(expected, filterCommand.toString());
+    }
+
+    /**
+     * Parses {@code userInput} into a {@code NameContainsKeyphrasePredicate}.
+     */
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/logic/commands/FilterCommandTest.java:86]: 👍

File: src/test/java/seedu/address/model/person/NameContainsKeyphrasePredicateTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        Person alice = new PersonBuilder().withName("Alice").withPhone("12345")
+                .withEmail("alice@email.com").withAddress("Main Street").build();
+
+        // Keyphrase match phone but does not match name
+        predicate = new NameContainsKeyphrasePredicate("12345");
+        assertFalse(predicate.test(alice));
+
+        // Keyphrase match email but does not match name
+        predicate = new NameContainsKeyphrasePredicate("alice@email.com");
+        assertFalse(predicate.test(alice));
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/model/person/NameContainsKeyphrasePredicateTest.java:81]: Thanks for adding tests!

File: src/test/java/seedu/address/model/person/NameContainsKeyphrasePredicateTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        // Keyphrase match address but does not match name
+        predicate = new NameContainsKeyphrasePredicate("Main");
+        assertFalse(predicate.test(alice));
+        predicate = new NameContainsKeyphrasePredicate("Street");
+        assertFalse(predicate.test(alice));
+
+    }
```
Comment (UdhayaShan1) [inline src/test/java/seedu/address/model/person/NameContainsKeyphrasePredicateTest.java:89]: Perhaps can add few tests for case-sensitive checks

**Review by UdhayaShan1 — APPROVED** [review]
Overall LGTM, can consider 
- a step further to throw `ParseException` if key phrase is not a alphanumeric phrase? Can do this in `parse` of `FilterCommand` utilising `ParserUtil`. This is a optional step to improve AB3 imo.
- throw new exception with appropriate message accordingly, can do this in 1.2b not urgent

**Review by appleraincoat — APPROVED** [review]
LGTM! Tests look great

**Pull request merged**